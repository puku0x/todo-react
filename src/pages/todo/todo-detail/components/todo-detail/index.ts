export * from './todo-detail.component';
