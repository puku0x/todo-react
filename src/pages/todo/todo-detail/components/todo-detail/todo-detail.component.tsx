import { memo } from 'react';
import { NavLink } from 'react-router-dom';

import { Todo } from '../../../../../models';

const datePipe = (date: number) => {
  return new Date(date).toISOString();
};

interface Props {
  isFetching?: boolean;
  todo: Todo | null;
}

export const TodoDetail = memo((props: Props) => {
  const { todo } = props;

  return (
    <>
      <NavLink to="/todos">Back to list</NavLink>
      <h2>todo-detail</h2>
      {todo ? (
        <>
          <p>
            <NavLink to={`/todos/${todo.id}/edit`}>Edit this todo</NavLink>
          </p>
          <table>
            <tbody>
              <tr>
                <td>title</td>
                <td>{todo.title}</td>
              </tr>
              <tr>
                <td>completed</td>
                <td>{`${todo.completed}`}</td>
              </tr>
              <tr>
                <td>createdAt</td>
                <td>{datePipe(todo.createdAt)}</td>
              </tr>
              <tr>
                <td>updatedAt</td>
                <td>{datePipe(todo.updatedAt)}</td>
              </tr>
            </tbody>
          </table>
        </>
      ) : (
        <>loading...</>
      )}
    </>
  );
});
