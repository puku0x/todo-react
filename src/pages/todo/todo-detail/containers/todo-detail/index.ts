export * from './todo-detail.container';
