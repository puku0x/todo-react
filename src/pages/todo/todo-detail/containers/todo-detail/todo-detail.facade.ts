import { AnyAction, ThunkDispatch, unwrapResult } from '@reduxjs/toolkit';
import { useCallback, useEffect } from 'react';
import { useDispatch, useSelector } from 'react-redux';

import {
  fetchTodo,
  isFetchingSelector,
  todoSelector,
} from '../../../../../store/todo';

export const useTodoDetailFacade = (arg: { id: string }) => {
  const { id } = arg;
  const dispatch = useDispatch<ThunkDispatch<unknown, undefined, AnyAction>>();
  const isFetching = useSelector(isFetchingSelector);
  const todo = useSelector(todoSelector);

  const fetch = useCallback(
    (arg: { id: string }) => {
      return dispatch(fetchTodo(arg)).then(unwrapResult);
    },
    [dispatch]
  );

  useEffect(() => {
    fetch({ id });
  }, [id, fetch]);

  return {
    isFetching,
    todo,
    fetch,
  } as const;
};
