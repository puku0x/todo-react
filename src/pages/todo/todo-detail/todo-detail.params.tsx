import { useParams } from 'react-router-dom';

interface RouterParams {
  id: string;
}

export const useTodoDetailParams = () => {
  const { id } = useParams<RouterParams>();

  return {
    id,
  } as const;
};
