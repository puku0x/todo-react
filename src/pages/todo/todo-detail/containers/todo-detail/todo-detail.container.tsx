import { memo } from 'react';

import { TodoDetail } from '../../components';
import { useTodoDetailFacade } from './todo-detail.facade';

interface Props {
  id: string;
}

export const TodoDetailContainer = memo((props: Props) => {
  const { id } = props;
  const { isFetching, todo } = useTodoDetailFacade({ id });

  return <TodoDetail isFetching={isFetching} todo={todo} />;
});
