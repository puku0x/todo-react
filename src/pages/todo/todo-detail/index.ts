import { lazy } from 'react';

export const TodoDetailPage = lazy(() =>
  import('./todo-detail.page').then((m) => ({ default: m.TodoDetailPage }))
);
