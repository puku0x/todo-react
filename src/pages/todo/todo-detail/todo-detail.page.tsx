import { memo } from 'react';
import { useParams } from 'react-router-dom';

import { TodoDetailContainer } from './containers';

interface RouterParams {
  id: string;
}

export const TodoDetailPage = memo(() => {
  const { id } = useParams<RouterParams>();

  return <TodoDetailContainer id={id} />;
});
