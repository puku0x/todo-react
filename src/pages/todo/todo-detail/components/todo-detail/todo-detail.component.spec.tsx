import { act, render } from '@testing-library/react';
import { Fragment } from 'react';
import { MemoryRouter } from 'react-router-dom';

import { createTodoMock } from '../../../../../models/testing';
import { TodoDetail } from './todo-detail.component';

describe('TodoDetail', () => {
  it('render', async () => {
    const todo = createTodoMock();
    let result = render(<Fragment />);
    await act(async () => {
      result = render(<TodoDetail todo={todo} />, { wrapper: MemoryRouter });
    });
    const { asFragment } = result;

    expect(asFragment()).toMatchSnapshot();
  });

  it('render with isFetching', async () => {
    let result = render(<Fragment />);
    await act(async () => {
      result = render(<TodoDetail isFetching todo={null} />, {
        wrapper: MemoryRouter,
      });
    });
    const { asFragment } = result;

    expect(asFragment()).toMatchSnapshot();
  });
});
