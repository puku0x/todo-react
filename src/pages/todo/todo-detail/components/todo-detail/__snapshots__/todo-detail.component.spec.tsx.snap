// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoDetail render 1`] = `
<DocumentFragment>
  <a
    href="/todos"
  >
    Back to list
  </a>
  <h2>
    todo-detail
  </h2>
  <p>
    <a
      href="/todos/1/edit"
    >
      Edit this todo
    </a>
  </p>
  <table>
    <tbody>
      <tr>
        <td>
          title
        </td>
        <td>
          title
        </td>
      </tr>
      <tr>
        <td>
          completed
        </td>
        <td>
          false
        </td>
      </tr>
      <tr>
        <td>
          createdAt
        </td>
        <td>
          1970-01-02T10:17:36.789Z
        </td>
      </tr>
      <tr>
        <td>
          updatedAt
        </td>
        <td>
          1970-01-02T10:17:36.789Z
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`TodoDetail render with isFetching 1`] = `
<DocumentFragment>
  <a
    href="/todos"
  >
    Back to list
  </a>
  <h2>
    todo-detail
  </h2>
  loading...
</DocumentFragment>
`;
